feat: support detailed `--help`

ikatyang opened the pull request:
main:

- Fixes #2826

side:

- replace `or` with `,`, e.g. `--help or -h` -> `-h, --help`
- add DidYouMean redirection for `--help <typo>` using leven

ikatyang commented on src/cli-constant.js:
```diff
@@ -136,9 +152,10 @@ const detailedOptions = normalizeDetailedOptions({
   },
   help: {
-    type: "boolean",
+    type: "option-name",
```
Don't know how to describe this type.

azz replied:
Flag?

ikatyang commented on tests_integration/__tests__/__snapshots__/early-exit.js.snap:
```diff
@@ -1,5 +1,77 @@
+  cli-override  | default config => config file => CLI options
+  file-override | default config => CLI options => config file
+  prefer-file   | default config => config file (if config file is found) or
+                  default config => CLI options (if no config file is found)
```
Should we remove these descriptions from `option.description`? I think they should belong to `choice.description`.

lydell commented on src/cli-constant.js:
```diff
@@ -241,9 +258,19 @@ const detailedOptions = normalizeDetailedOptions({
+        value: "all",
+        description: dedent(`
+          Trailing commas wherever possible (including function arguments).
+          This requires node 8 or a transform.
```
I understand that you copied this from the readme, but I don't think we should include information that will be quickly outdated here. We can change the readme at any time, but changing the CLI description requires a new version to be release. Both Firefox and Chrome supports trailing function argument commas (not sure about other browsers). I think we can leave out the "This requires node 8 or a transform." sentence (and update the readme at some point).

ikatyang replied:
👍

lydell commented on src/cli-constant.js:
```diff
@@ -148,10 +148,10 @@ const detailedOptions = normalizeDetailedOptions({
   },
   help: {
-    type: "option-name",
+    type: "flag",
```
Nice! 👍

lydell commented on src/cli-util.js:
```diff
@@ -404,6 +410,77 @@ function createOptionUsageType(option) {
+    }
+  }
+
+  const option = options[optionIndex];
```
Perhaps extract the above into a function so that we can do `const option = fooGetOption(optionName);` here?

lydell commented on src/cli-util.js:
```diff
@@ -404,6 +410,77 @@ function createOptionUsageType(option) {
+    if (suggestionOptionIndex !== -1) {
+      const suggestionOptionName = options[suggestionOptionIndex].name;
+      console.warn(
+        `Unexpected option name "${optionName}", did you mean "${suggestionOptionName}"?\n`
```
I think "Unknown" is better than "Unexpected".

lydell commented on src/cli-util.js:
```diff
@@ -404,6 +410,77 @@ function createOptionUsageType(option) {
+      optionName = suggestionOptionName;
+      optionIndex = suggestionOptionIndex;
+    } else {
+      throw new Error(`Unexpected option name "${optionName}"`);
```
Same here.

lydell commented on src/cli-util.js:
```diff
@@ -404,6 +410,77 @@ function createOptionUsageType(option) {
+            )
+          );
+          return `\n\nValid options:\n\n${choiceUsages.join("\n")}`;
+        })();
```
Would it be possible to extract this function into ... a standalone function?

lydell commented on src/cli-util.js:
```diff
@@ -404,6 +410,77 @@ function createOptionUsageType(option) {
+function kebabToTitle(str) {
+  return str.replace(
+    /(^|-)([a-z])/g,
+    (_, prefix, char) => " ".repeat(prefix.length) + char.toUpperCase()
```
The `.repeat()` thing is almost a bit too clever to me :)

Do we even need `kebabToTile` though? Can't we just show `--bracket-spacing` instead of `Bracket Spacing` in the header?

lydell commented on tests_integration/__tests__/__snapshots__/early-exit.js.snap:
```diff
@@ -1,5 +1,72 @@
+  typescript   undocumented
+  postcss      undocumented
+  json         undocumented
+  graphql      undocumented
```
I think it's better to omit the choice descriptions if they are missing instead of defaulting to "undocumented".

lydell commented on tests_integration/__tests__/__snapshots__/early-exit.js.snap:
```diff
@@ -1,5 +1,72 @@
+  es5    Trailing commas where valid in ES5 (objects, arrays, etc.)
+  all    Trailing commas wherever possible (including function arguments).
+
+Default: none
```
It's a bit unfortunate that it says "Defaults to none." above and then we repeat "Default: none" again here. Is there anything we can do about that?

lydell commented on src/cli-constant.js:
```diff
@@ -97,13 +97,25 @@ const detailedOptions = normalizeDetailedOptions({
-      file-override | default config => CLI options => config file
-      prefer-file   | default config => config file (if config file is found) or
-                      default config => CLI options (if no config file is found)
+      Define in which order config files and CLI options should be evaluated.
```
Is this description supposed to be on two lines?

lydell commented on src/cli-constant.js:
```diff
@@ -136,9 +148,10 @@ const detailedOptions = normalizeDetailedOptions({
     alias: "h",
-    description: "Show help."
+    description:
+      "Show CLI usages, or detailed option usage if flag name present."
```
What about this?

> Show CLI usage, or details about the given flag.

azz commented on tests_integration/__tests__/early-exit.js:
```diff
@@ -17,6 +17,57 @@ test("show usage with --help", () => {
   expect(result.status).toEqual(0);
 });
 
+test("show detailed usage with --help trailing-comma", () => {
```
Thoughts on iterating over all of the options:

```js
const options = require("../../src/cli-constant").detailedOptions;

options.forEach(option => {
  test(`show detailed usage with --help ${option.name}`, () => {
    // ...
  });
});
```

The advantage is that when we add a new option we're forced to look at how it will show in help.

ikatyang commented:
Addressed all the feedbacks and dynamically generate `defaults to ...` description for usage overview so as to avoid duplicate `default ...` description in detailed usages.

lydell commented:
Here is a case we must handle:

```
$ ./bin/prettier.js --help --help
/home/lydell/forks/prettier/node_modules/leven/index.js:35
	while (aLen > 0 && (a.charCodeAt(~-aLen) === b.charCodeAt(~-bLen))) {
	                      ^

TypeError: a.charCodeAt is not a function
    at module.exports (/home/lydell/forks/prettier/node_modules/leven/index.js:35:24)
    at optionNames.findIndex.currentOptionName (/home/lydell/forks/prettier/src/cli-util.js:431:26)
    at Array.findIndex (native)
    at getOptionWithLevenSuggestion (/home/lydell/forks/prettier/src/cli-util.js:430:45)
    at Object.createDetailedUsage (/home/lydell/forks/prettier/src/cli-util.js:461:18)
    at Object.run (/home/lydell/forks/prettier/src/cli.js:27:16)
    at Object.<anonymous> (/home/lydell/forks/prettier/bin/prettier.js:5:23)
    at Module._compile (module.js:573:30)
    at Object.Module._extensions..js (module.js:584:10)
    at Module.load (module.js:507:32)
```

azz commented:
Should we catch this error?

```
azz at macOS in ~/code/prettier on 79baf1e5 [$]
$ node ./bin/prettier.js -h no-use-tabs
/Users/azz/code/prettier/src/cli-util.js:443
  throw new Error(`Unknown option name "${optionName}"`);
  ^

Error: Unknown option name "no-use-tabs"
    at getOptionWithLevenSuggestion (/Users/azz/code/prettier/src/cli-util.js:443:9)
    at Object.createDetailedUsage (/Users/azz/code/prettier/src/cli-util.js:461:18)
    at Object.run (/Users/azz/code/prettier/src/cli.js:29:18)
    at Object.<anonymous> (/Users/azz/code/prettier/bin/prettier.js:5:23)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
```

ikatyang commented:
The error suggestion is based on [Levenshtein Distance](https://en.wikipedia.org/wiki/Levenshtein_distance) so that we could:

- relax the threshold (current: leven dist < 3)
  - `Did you mean 'use-tabs'?`
- or add `oppositeDescription` for `use-tabs`
  - show description for `no-use-tabs`

azz commented:
Misleading example, same happens if you do `--help blargh`

ikatyang commented:
Since `--help --other-flag` is treated as `--help`, so I removed the special case of `--help --help`.

New rules:

- print `--help <suggested>` message for `--help <typo>` (leven dist < 3)
- print `--help help` message for `--help <unknown>`
- print `--help` message for `--help --other-flag`

lydell commented:
We need to handle boolean options that are usually only used with `--no-` prefix, but which are valid both ways, such as `--semi` and `--color`.

```
$ ./bin/prettier.js --help semi
Unknown option name "semi"

-h, --help <flag>

  Show CLI usage, or details about the given flag.
  Example: --help write
```

We could start with just showing the `--no-` version:

```
$ ./bin/prettier.js --help semi
--no-semi

  Do not print semicolons, except at the beginning of lines which may need them.
```

But I don't know if that's the best solution.

ikatyang commented:
I'd prefer adding their own description (e.g. `semi`, `color`, etc.) as it looks more complete, but should we only show them in their own `--help <flag>`? as it looks a little weird in the `--help` with both of `no-option` and `option` existed simultaneously:

![image](https://user-images.githubusercontent.com/8341033/30649909-1dd16bb0-9e54-11e7-892d-d046a104f886.png)

![image](https://user-images.githubusercontent.com/8341033/30650208-e74d3a0a-9e54-11e7-969c-dc476d22e901.png)

lydell commented:
- We should not show `--semi`, `--color` etc. in `--help`.
- I don't really care what `--help semi` does as long as it is not a lie (which "Unknown option semi" is).
